File: src/kame_project/src/tour.cpp
#include "ros/ros.h"
#include "move_base_msgs/MoveBaseAction.h"
#include "actionlib/client/simple_action_client.h"
#include "geometry_msgs/PoseWithCovarianceStamped.h" 

// goal estimates
double XGoal1 = 2.53566564977;
double YGoal1 = -0.773420049965;

double XGoal2 = 4.49860523334;
double YGoal2 = 0.348575832826;

double XGoal3 = 2.33846501217;
double YGoal3 = 2.30097446639;

double XGoal4 = -0.0342952696324;
double YGoal4 = 1.04674247832;

bool moveToGoal(double X, double Y) {
    actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> ac("move_base", true);

    // wait for the action server to come up
    while(!ac.waitForServer(ros::Duration(5.0))){
       ROS_INFO("Waiting for the move_base action server to come up");
    }

    move_base_msgs::MoveBaseGoal goal;

    // set up the frame parameters
    goal.target_pose.header.frame_id = "map";
    goal.target_pose.header.stamp = ros::Time::now();

   // moving towards the goal

    goal.target_pose.pose.position.x =  X;
    goal.target_pose.pose.position.y =  Y;
    goal.target_pose.pose.position.z =  0.0;
    goal.target_pose.pose.orientation.x = 0.0;
    goal.target_pose.pose.orientation.y = 0.0;
    goal.target_pose.pose.orientation.z = 0.0;
    goal.target_pose.pose.orientation.w = 1.0;

    ROS_INFO("Sending goal location ...");
    ac.sendGoal(goal);

    ac.waitForResult();

    if(ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED){
       ROS_INFO("You have reached the destination");
       return true;
    }
    else{
       ROS_INFO("The robot failed to reach the destination");
    }
    return false;
}

void localize(const geometry_msgs::PoseWithCovarianceStamped::ConstPtr &msg) {
    double x = msg->pose.pose.position.x;
    double y = msg->pose.pose.position.y;
    
    ROS_INFO("Position: x = %f, y = %f", x, y);

}

int main (int argc, char **argv) {
    ros::init(argc, argv, "tour");
    ros::NodeHandle n;
       
    // wait for everything to be set up first
    ros::Duration(5.0).sleep();
    
    ros::Subscriber detect_sub = n.subscribe("amcl_pose", 10, localize);
    
    ros::spin();
    
    return 0;
}

File: src/kame_project/scripts/tour.py
#!/usr/bin/env python

import rospy
import actionlib
from actionlib_msgs.msg import *
from geometry_msgs.msg import PoseWithCovarianceStamped, Point
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
import itertools
import numpy as np

# starting point of robot: updated through subscriber
position = [0, 0]


# gets the shortest path from predefined points and a origin
# from the robot's current position
def get_shortest_path(x,y):
    # Points and origin values
    origin = [x,y]                           # UPDATE TO CORRECT ORIGIN
    point_1 = [2.53566564977, -0.773420049965]
    point_2 = [4.49860523334, 0.348575832826]
    point_3 = [2.33846501217, 2.30097446639]
    point_4 = [-0.0342952696324, 1.04674247832]
    points = [point_1, point_2, point_3, point_4]

    # Now get all possible paths by getting permutations
    paths = list(itertools.permutations(points))

    # Iterate through each path
    distances = []
    for path in paths:
        dist = 0
        current_origin = origin
        for point in path:
            dist += np.linalg.norm(np.array(point)-np.array(current_origin))
            current_origin = point

        distances.append(dist)

    # Get the index of the shortest path
    shortest_index = distances.index(min(distances))

    # Get shortest path from paths using shortest_index
    shortest_path = paths[shortest_index]

    return shortest_path

# updates the current position of the turtlebot relative to map
def localize(msg):
    position[0] = msg.pose.pose.position.x
    position[1] = msg.pose.pose.position.y
    # rospy.loginfo('position updated: x = %f, y = %f', position[0], position[1])

def main():
    # setup the node and subscriber.
    # NOTE: subscriber runs in its own thread so no need for rospy.spin()
    rospy.init_node('tour', anonymous=True)
    rospy.Subscriber('amcl_pose', PoseWithCovarianceStamped, localize)
    
    # Loops for user input
    while not rospy.is_shutdown():
        user = raw_input('Start the tour? (y/n): ').lower()
        # the yes option
        if user == 'y':
            rospy.loginfo('Thank you for joining us today. Pleas wait for us to get the tour ready :o')
            path = get_shortest_path(position[0], position[1])
            rospy.sleep(1.0)
                        
            # set up for sending goals
            client = actionlib.SimpleActionClient('move_base', MoveBaseAction)
            client.wait_for_server()
            
            rospy.loginfo('Thank you for waiting. We will now begin our tour :)')
            for point in path:
                goal = MoveBaseGoal()

                #set up the frame parameters
                goal.target_pose.header.frame_id = "map"
                goal.target_pose.header.stamp = rospy.Time.now()

                # moving towards the current point in path
                goal.target_pose.pose.position =  Point(point[0], point[1],0)
                goal.target_pose.pose.orientation.x = 0.0
                goal.target_pose.pose.orientation.y = 0.0
                goal.target_pose.pose.orientation.z = 0.0
                goal.target_pose.pose.orientation.w = 1.0
                
                rospy.loginfo('Sending goal location ...')
                
                # send path and wait for result
                client.send_goal(goal)                
                client.wait_for_result()
                
                # continue only if goal successfully reached otherwise end the tour
                if(client.get_state() ==  GoalStatus.SUCCEEDED):
                    rospy.loginfo('Here we have [insert location name] which is [insert location desc] :P')

                else:
                    rospy.loginfo('The robot failed to reach the destination')
                    break
                    
            rospy.loginfo('This concludes our tour :D')
                
        # the no option
        else:
            rospy.loginfo('Have a nice day (^_^)')
    


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass
        
